frame/evm/precompile/contracts/kton/src/lib.rs

```diff
@@ -0,0 +1,189 @@
+// This file is part of Darwinia.
+//
+// Copyright (C) 2018-2021 Darwinia Network
+// SPDX-License-Identifier: GPL-3.0
+//
+// Darwinia is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Darwinia is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Darwinia. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::traits::{Currency, ExistenceRequirement};
+use sha3::Digest;
+use sp_core::{H160, U256};
+use sp_runtime::traits::UniqueSaturatedInto;
+use sp_std::marker::PhantomData;
+use sp_std::prelude::*;
+use sp_std::vec::Vec;
+
+use core::str::FromStr;
+use darwinia_evm::{AddressMapping, Runner, Trait};
+use darwinia_evm_primitives::Precompile;
+use ethabi::{Function, Param, ParamType, Token};
+use evm::{Context, ExitError, ExitSucceed};
+use sp_std::borrow::ToOwned;
+
+const DEPOSIT_FUNC: &[u8] = b"deposit(address,uint256)";
+const KTON_PRECOMPILE: &str = "0000000000000000000000000000000000000016";
+/// Kton Precompile Contract
+///
+/// The contract address: 0000000000000000000000000000000000000016
+pub struct Kton<T: Trait> {
+	_maker: PhantomData<T>,
+}
+
+impl<T: Trait> Precompile for Kton<T> {
+	// The execute process of Kton precompile, the input data consists of four parts:
+	// 1. Wkton solidity contract deployed address(20 bytes)
+	// 2. The function selector(4 bytes)
+	// 3. The deposit kton evm address(20 bytes)
+	// 4. The deposit kton value(32 bytes)
+	fn execute(
+		input: &[u8],
+		target_limit: Option<usize>,
+		context: &Context,
+	) -> core::result::Result<(ExitSucceed, Vec<u8>, usize), ExitError> {
+		let helper = U256::from(10)
+			.checked_pow(U256::from(9))
+			.unwrap_or(U256::MAX);
+		let deposit_func = &sha3::Keccak256::digest(&DEPOSIT_FUNC)[0..4];
+		// decode input
+		let input = InputData::decode(input)?;
+
+		let con_caller = T::AddressMapping::into_account_id(context.caller);
+		if hex::encode(input.func_selector) == hex::encode(&deposit_func) {
+			// 1. Transfer kton from sender to kton erc20 contract
+			let wkton_account_id = T::AddressMapping::into_account_id(input.wkton_address);
```

Comment:
what's the difference between `input.address` and `wkton_account_id `?

Reply:
Let's not use two addresses here, input.address can do redirection to transfer the kton value to any other address he would like.

In the case of WKTON, `wkton_account_id` is encoded as part of the `call_data`, it is `to`(WKTON here) to decide how to parse the `call_data` and how to use it. We do not need to specify `wkton_account_id` in precompile input.

If the target `to` address is not WKTON, there could be no such redirection address.

@hujw77 Would you please help provide a WKTON implementation for test case purpose?

Reply:
https://github.com/evolutionlandorg/evo-deploy/blob/main/src/wrapper/WKTON.sol

Reply:
https://github.com/evolutionlandorg/evo-deploy/blob/main/src/wrapper/WKTON.sol
@HackFisher Need to be reviewed too.

Reply:
@hujw77 WKTON is better to move to darwinia repository, can also be shared as common used contract for other applications.

https://github.com/evolutionlandorg/evo-deploy/blob/f340fcf67219f77ee788e880099fd378920198dc/src/wrapper/WKTON.sol#L21

Need to check the decimal convert, the kton_value from KTON Precompile is 9 now.

Need to check whether need an extra `receiver` address in `function deposit(address from, uint256 value)`, with or without it, both ok for me. Now I prefer be simpler, aka. without `receiver`wkton_account_id, as it is.